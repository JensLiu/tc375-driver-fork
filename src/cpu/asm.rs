/* From ILLD file: 
*  file IfxCpu.h
*  brief CPU  basic functionality
*  ingroup IfxLld_Cpu
*  version iLLD_1_0_1_12_0
*  copyright Copyright (c) 2019 Infineon Technologies AG. All rights reserved.
*/

// global interrupts enable
#[inline]
pub fn enable_interrupts() {
    #[cfg(target_arch = "tricore")]
    unsafe {
        core::arch::asm!("enable");
    }
}

// TODO (annabo) // to be tested
/* ILLD FUNCTION 
 * IFX_INLINE boolean IfxCpu_disableInterrupts(void)
 * {
 *     boolean enabled;
 *     enabled = IfxCpu_areInterruptsEnabled();
 *     __disable();
 *     __nop();
 *     return enabled;
 * }
 * */

// global interrupts disable
#[inline]
pub fn disable_interrupts() {
    #[cfg(target_arch = "tricore")]
    unsafe {
        core::arch::asm!("disable");
    }
}

/** \brief FE1C, CPUx Core Identification Register */
#[allow(dead_code)]
const CPU_CORE_ID: u32 = 0xFE1C;

/**
 * Read CPU core id.
 * IFX_INLINE IfxCpu_Id IfxCpu_getCoreId(void)
 * {
 *     Ifx_CPU_CORE_ID reg;
 *     reg.U = __mfcr(CPU_CORE_ID);
 *     return (IfxCpu_Id)reg.B.CORE_ID;
 * }
*/

#[inline]
#[cfg(target_arch = "tricore")]
pub fn read_cpu_core_id() -> u32 {
    #[allow(unused_assignments)]
    let value: u32;
    unsafe {
        core::arch::asm!("mfcr {0}, 0xFE1C", out(reg32) value);
    }
    value
}

#[inline]
#[cfg(not(target_arch = "tricore"))]
pub fn read_cpu_core_id() -> u32 {
    0
}
