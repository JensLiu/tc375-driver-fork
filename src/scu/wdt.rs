use crate::cpu::asm::read_cpu_core_id;
use tc37x_pac as pac;

#[inline]
pub fn get_cpu_watchdog_password() -> u16 {
    let core_id = read_cpu_core_id();
    let password = match core_id {
        0 => unsafe { pac::SCU.wdtcpu0con0().read() }.pw().get(),
        1 => unsafe { pac::SCU.wdtcpu1con0().read() }.pw().get(),
        2 => unsafe { pac::SCU.wdtcpu2con0().read() }.pw().get(),
        _ => unreachable!(),
    };

    password ^ 0x003F
}

#[inline]
pub fn get_safety_watchdog_password() -> u16 {
    let password = unsafe { pac::SCU.wdtscon0().read() }.pw().get();
    password ^ 0x003F
}

#[inline]
unsafe fn get_wdt_con0(core_id: u8) -> pac::Reg<pac::scu::Wdtcpu0Con0, pac::RW> {
    // unsafe cast to get the valid SCU WDT based on the core id
    let off: *mut u8 = unsafe { core::mem::transmute(pac::SCU.wdtcpu0con0()) };
    let off = unsafe { off.add(core::mem::size_of::<u32>() * 3 * core_id as usize) };
    unsafe { core::mem::transmute(off) }
}

#[inline]
pub fn clear_cpu_endinit_inline(password: u16) {
    let core_id = read_cpu_core_id();
    let con0 = unsafe { get_wdt_con0(core_id as _) };

    if unsafe { con0.read() }.lck().get() {
        let rel = unsafe { con0.read() }.rel().get();
        let data = pac::scu::Wdtcpu0Con0::default()
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        unsafe { con0.write(data) };
    }

    let rel = unsafe { con0.read() }.rel().get();
    let data = pac::scu::Wdtcpu0Con0::default()
        .endinit()
        .set(false)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);
    unsafe { con0.write(data) };

    #[cfg(tricore_arch = "tricore")]
    while unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub fn set_cpu_endinit_inline(_password: u16) {
    // TODO
}

#[inline]
unsafe fn get_wdt_con0_cpu2() -> pac::Reg<pac::scu::Wdtcpu2Con0, pac::RW> {
    // unsafe cast to get the valid SCU WDT based on the core id
    let off: *mut u8 = unsafe { core::mem::transmute(pac::SCU.wdtcpu0con0()) };
    let off = unsafe { off.add(core::mem::size_of::<u32>() * 3 * 2usize) };
    unsafe { core::mem::transmute(off) }
}
#[inline]
unsafe fn get_wdt_con0_cpu1() -> pac::Reg<pac::scu::Wdtcpu1Con0, pac::RW> {
    // unsafe cast to get the valid SCU WDT based on the core id
    let off: *mut u8 = unsafe { core::mem::transmute(pac::SCU.wdtcpu0con0()) };
    let off = unsafe { off.add(core::mem::size_of::<u32>() * 3) };
    unsafe { core::mem::transmute(off) }
}
#[inline]
unsafe fn get_wdt_con0_cpu0() -> pac::Reg<pac::scu::Wdtcpu0Con0, pac::RW> {
    // unsafe cast to get the valid SCU WDT based on the core id
    let off: *mut u8 = unsafe { core::mem::transmute(pac::SCU.wdtcpu0con0()) };
    unsafe { core::mem::transmute(off) }
}

#[inline]
pub fn set_cpu_endinit_inline_cpu0(password: u16) {
    let con0 = unsafe { get_wdt_con0_cpu0() };

    if unsafe { con0.read() }.lck().get() {
        let rel = unsafe { con0.read() }.rel().get();
        let data = pac::scu::Wdtcpu0Con0::default()
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        unsafe { con0.write(data) };
    }

    let rel = unsafe { con0.read() }.rel().get();
    let data = pac::scu::Wdtcpu0Con0::default()
        .endinit()
        .set(true)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);

    unsafe { con0.write(data) };

    #[cfg(tricore_arch = "tricore")]
    while !unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub fn clear_safety_endinit_inline(password: u16) {
    let con0 = pac::SCU.wdtscon0();

    if unsafe { con0.read() }.lck().get() {
        unsafe { con0.modify(|r| r.endinit().set(true).lck().set(false).pw().set(password)) };
    }
    unsafe { con0.modify(|r| r.endinit().set(false).lck().set(true).pw().set(password)) }

    #[cfg(tricore_arch = "tricore")]
    while unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub fn set_cpu_endinit_inline_cpu1(password: u16) {
    let con0 = unsafe { get_wdt_con0_cpu1() };

    if unsafe { con0.read() }.lck().get() {
        let rel = unsafe { con0.read() }.rel().get();
        let data = pac::scu::Wdtcpu1Con0::default()
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        unsafe { con0.write(data) };
    }

    let rel = unsafe { con0.read() }.rel().get();
    let data = pac::scu::Wdtcpu1Con0::default()
        .endinit()
        .set(true)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);

    unsafe { con0.write(data) };

    #[cfg(tricore_arch = "tricore")]
    while !unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub fn set_cpu_endinit_inline_cpu2(password: u16) {
    let con0 = unsafe { get_wdt_con0_cpu2() };

    if unsafe { con0.read() }.lck().get() {
        let rel = unsafe { con0.read() }.rel().get();
        let data = pac::scu::Wdtcpu2Con0::default()
            .endinit()
            .set(true)
            .lck()
            .set(false)
            .pw()
            .set(password)
            .rel()
            .set(rel);
        unsafe { con0.write(data) };
    }

    let rel = unsafe { con0.read() }.rel().get();
    let data = pac::scu::Wdtcpu2Con0::default()
        .endinit()
        .set(true)
        .lck()
        .set(true)
        .pw()
        .set(password)
        .rel()
        .set(rel);

    unsafe { con0.write(data) };

    #[cfg(tricore_arch = "tricore")]
    while !unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub fn clear_safety_endinit_inline_cpu0(password: u16) {
    let con0 = pac::SCU.wdtscon0();

    if unsafe { con0.read() }.lck().get() {
        unsafe { con0.modify(|r| r.endinit().set(true).lck().set(false).pw().set(password)) };
    }
    unsafe { con0.modify(|r| r.endinit().set(false).lck().set(true).pw().set(password)) }

    #[cfg(tricore_arch = "tricore")]
    while unsafe { con0.read() }.endinit().get() {}
}

#[inline]
pub fn set_safety_endinit_inline(password: u16) {
    let con0 = pac::SCU.wdtscon0();

    if unsafe { con0.read() }.lck().get() {
        unsafe { con0.modify(|r| r.endinit().set(true).lck().set(false).pw().set(password)) };
    }

    unsafe { con0.modify(|r| r.endinit().set(true).lck().set(true).pw().set(password)) }
    #[cfg(tricore_arch = "tricore")]
    while !unsafe { con0.read() }.endinit().get() {}
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    pub fn test_get_wdt_con0() {
        let pwd = unsafe { get_cpu_watchdog_password() };
        assert_eq!(pwd, 0x3F);
    }
}
