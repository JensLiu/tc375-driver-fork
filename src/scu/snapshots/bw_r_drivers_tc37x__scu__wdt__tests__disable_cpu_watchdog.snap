---
source: src/scu/wdt.rs
assertion_line: 246
expression: report.take_log()
---
r    0xF003624C 04 0x00000003
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FE
r    0xF0036250 04 0x00000000
w    0xF0036250 04 0x00000008
r    0xF003624C 04 0x00000002
w    0xF003624C 04 0x000000FD
w    0xF003624C 04 0x000000FF
